llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -610,6 +658,48 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
       // unknown users
     }
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = cast<Value>(U);
+    if (IsRecursive && CB && PtrArg) {
```

Comment:
This IsRecursive check doesn't look right: it checks whether the function contains a recursive call somewhere, but it doesn't check whether this particular call is recursive.

Reply:
I had added a check to check if a function is self recursive, and not a chain of recursion (pr42028-recursion.ll). Does that seem okay?

Reply:
pushed a patch addressing both the comments.

Reply:
You're still not checking if the call in question is recursive; a function can contain both recursive and non-recursive calls.

Reply:
> You're still not checking if the call in question is recursive; a function can contain both recursive and non-recursive calls.

Oh so, just a check if CB.getCalledFunction() == CB.getFunction()! Done and pushed.

Reply:
Yes, that looks right.  (And with that, you don't really need the IsSelfRecursive check here.)

Reply:
Should I still pass IsSelfRecursive ? seems it might come in handy, because IsRecursive is quite generic, as it just uses SCC.size()